fuzzers/main.py

```diff
@@ -0,0 +1,450 @@
+    *,
+    exit_event: threading.Event,
+    path: pathlib.Path,
+    local: pathlib.Path,
+    remote: str,
+    bucket: str,
+    log_file: typing.IO[str],
+    ignore_deletions: bool,
+    # Without Any here mypy doesn't let us use lambdas. The result is ignored anyway.
+    on_uploaded_item: typing.Callable[[str], typing.Any],
+    on_deleted_item: typing.Callable[[str], typing.Any],
+) -> None:
+    """
+    Setup an inotify thread that will automatically upload changes to GCS
+
+    `exit_event` will stop the thread when set
+    The contents of `local/path` will be propagated to `remote/path` in `bucket`
+    Logs will be sent to `log_file`.
+    If `ignore_deletions` is set, deletions will not be propagated.
+    Upon successful uploading of deletion or an item, resp. `on_uploaded_item` and
+    `on_deleted_item` will be called.
+    """
+    # pylint: disable=line-too-long
+    print(f"Setting up inotify watch to auto-upload changes to {local / path} to GCS {remote}/{path}/")
+    log_file.write(f"Setting up inotify watch to auto-upload changes to {local / path} to GCS {remote}/{path}/\n")
+    log_file.flush()
+    # pylint: enable=line-too-long
+
+    threading.Thread(
+        daemon = True,
+        target = uploader_thread,
+        args = (path, local, remote, bucket, exit_event, log_file),
+    ).start()
+
+# TODO: replace Any here with something more precise
+THREAD_EXCEPTION: typing.Optional[typing.Any] = None
+EXCEPTION_HAPPENED_IN_THREAD = threading.Event()
+
+def main() -> None:
+    """Main function"""
+    # Make sure to cleanup upon ctrl-c or upon any exception in a thread
+    def new_excepthook(args: typing.Any) -> None:
+        global THREAD_EXCEPTION, EXCEPTION_HAPPENED_IN_THREAD
+        print(f"!! Caught exception in thread: {args}")
+        THREAD_EXCEPTION = args
+        EXCEPTION_HAPPENED_IN_THREAD.set()
+    threading.excepthook = new_excepthook
+
+    try:
+        # Start the system
+        gcs = connect_to_gcs()
+        start_fuzzing(gcs)
+
+        # Start the metrics server
+        prometheus_client.start_http_server(METRICS_PORT)
+
+        # And listen for the commands that might come up
+        threading.Thread(daemon = True, target = listen_for_commands, args = (gcs,)).start()
+
+        # Run until an exception forces us to stop
+        print("Startup complete, will run forever now")
+        while not EXCEPTION_HAPPENED_IN_THREAD.wait(timeout=AUTO_REFRESH_INTERVAL_SECS):
+            # Restart the fuzzers, thus getting latest commit and corpus, at every refresh interval
+            stop_fuzzers()
+            start_fuzzing(gcs)
+
+        # Finally, proxy the exception so it gets detected and acted upon by a human
+        exc_info = typing.cast(typing.Any, THREAD_EXCEPTION) # TODO: remove Any here
+        print(f"!! Got exception from thread in main thread {exc_info}")
+        sys.excepthook(exc_info.exc_type, exc_info.exc_value, exc_info.exc_traceback)
+        raise exc_info.exc_value
```

Comment:
I still don’t really like it. Maybe it’s a bit pedantic but we’ll print `exc_info` three times. How about `raise RuntimeError('Got exception in thread') from exc_info.exc_value`? I think that should work.